pkg/logql/log/labels.go

```diff
@@ -736,47 +738,53 @@ Outer:
 				}
 			}
 		}
-		for _, l := range b.base {
+
+		// HACK: we use Validate here so we can terminate early once we've found the desired label.
+		// (Range doesn't support terminating the loop early.)
+		_ = b.base.Validate(func(l labels.Label) error {
 			if g == l.Name {
 				b.buf = append(b.buf, l)
-				continue Outer
+
+				return errTerminateLoopEarly
 			}
-		}
+
+			return nil
+		})
 	}
 	return b.toUncategorizedResult(b.buf)
 }
 
 func (b *LabelsBuilder) withoutResult() LabelsResult {
 	if b.buf == nil {
-		size := len(b.base) + b.sizeAdd() - len(b.del) - len(b.groups)
+		size := b.base.Len() + b.sizeAdd() - len(b.del) - len(b.groups)
 		if size < 0 {
 			size = 0
 		}
-		b.buf = make(labels.Labels, 0, size)
+		b.buf = make([]labels.Label, 0, size)
 	} else {
 		b.buf = b.buf[:0]
```

Comment:
`b.buf` should really be a special builder or we move `LabelsBuilder` all together to avoid nesting mutliple builders.

Reply:
Not sure I'm following your comment here sorry - could you elaborate?